java/src/jmri/jmrix/pi/RaspberryPiSensorManager.java

```diff
@@ -50,15 +52,41 @@ public boolean isPullResistanceConfigurable(){
        return true;
     }
     
+   /**
+     * Require address portion of SystemName to either start with ":" or be numberic.
+     * {@inheritDoc} 
+     */
+    @Nonnull
+    @Override
+    public String createSystemName(@Nonnull String curAddress, @Nonnull String prefix) throws JmriException {
+        if (curAddress.substring(0,1).equals (":")) {
+            return prefix + typeLetter() + curAddress;
+        } else {
+            return prefix + typeLetter() + checkNumeric(curAddress);
+        }
+    }
+    
     /**
-     * Validates to Integer Format 0-999 with valid prefix.
-     * eg. PS0 to PS999
+     * Validates to either ":xxx..." or Integer Format 0-999 with valid prefix.
+     * eg. PT0 to PT999, PT:MCP23017:1:32:0
      * {@inheritDoc}
      */
     @Override
     @Nonnull
-    public String validateSystemNameFormat(@Nonnull String name, @Nonnull java.util.Locale locale) throws jmri.NamedBean.BadSystemNameException {
-        return this.validateIntegerSystemNameFormat(name, 0, 999, locale);
+    public String validateSystemNameFormat(@Nonnull String name, @Nonnull java.util.Locale locale) throws jmri.NamedBean.BadSystemNameException, IllegalArgumentException {
+        if (name.length() < 3) {
+            throw new IllegalArgumentException();
+        }
+        if (name.substring (2,3).equals (":")) {
+            return name;
+        } else {
+            return this.validateIntegerSystemNameFormat(name, 0, 999, locale);
+        }
     }
```

Comment:
Please do not throw IllegalArgumentException in this method. If the input string is bad, it is a bad system name, so BadSystemNameException should be thrown.

The code doesn't work as expected. The connection prefix might be a single letter, for example "L", or a letter followed by a number, for example "L12". So you need to use the method Manager.getSystemNamePrefix() to get the length of the system prefix and system letter:

`if (name.length() < getSystemNamePrefix().length+1) {`

---

If I understand it correctly, a system name for MCP23017 needs to be on the format:
`<system name prefix>:MCP23017:<number1>:<number2>:<number3>`

But you don't check that. The only thing you check is that the string has a colon. That's not enough. The purpose of this method is to protect the user from creating invalid system names. If an exact format is required, this method should check that the system name follows that format.

My recommendation is that you do something like this:

```
public String validateSystemNameFormat(@Nonnull String name, @Nonnull java.util.Locale locale) throws jmri.NamedBean.BadSystemNameException {
    if ((name.length() > getSystemNamePrefix().length()+":MCP23017:".length())
            && name.substring(getSystemNamePrefix().length(), getSystemNamePrefix().length()+":MCP23017:".length()).equals(":MCP23017:")) {

        String data = name.substring(getSystemNamePrefix().length() + ":MCP23017:".length());
        String parts[] = data.split(":");
        if (parts.length != 3) throw new BadSystemNameException();

        try {
            int bus = Integer.parseInt(parts[0]);
            int channel = Integer.parseInt(parts[1]);
            int pin = Integer.parseInt(parts[2]);
            
            if (bus < 0 || bus > 255) throw new BadSystemNameException();
            if (channel < 32 || channel > 39) throw new BadSystemNameException();
            if (pin < 0 || pin > 15) throw new BadSystemNameException();
            
            return name;
        } catch (NumberFormatException ignore) {
            throw new BadSystemNameException();
        }
    }

    return this.validateIntegerSystemNameFormat(name, 0, 999, locale);
}

```

Reply:
OK - Change to throw BadSystemNameException
OK - Use getSystemNamePrefix
No - Can't check for a three part address here.

If the MCP23017 extension was the only one possible, I could check for the three part address.  But a major goal of my code is to make these extensions user-definable.  The only requirement I want to impose (here) is that the string begin with a colon.  The user definable code (represented by ProvisionMCP23017) has code to validate the system name...for an MCP23017 device.  At the time that RaspberryPiSensorManager is running we don't know what type of device is involved.

I could add a validateSystemNameFormat entry to RaspberryPiGpioExFactory, which could invoke the validateSystemNameFormat entry for the appropriate device (ie, ProvisionMCP23017)...  I'll try making these changes and see how it comes out.

I must say that the system name handling seems arcane and fraught with less-than-ideal testing.  In this case, for example, the code suggests that the name is valid if the pin number is between 0 and 999.  Since the latest Raspberry Pi's have a total of 40 pins...and not all of them are usable...I'm not sure why we aren't checking for a much smaller range, or even figuring out what the maximum actually is.  I think I saw somewhere a reference to "9999" as well.

Reply:
> But a major goal of my code is to make these extensions user-definable.

What do you mean with that? Lets assume I (as an ordinary user without knowing how to change the source code) purchase the device `XYZ1234` and connects it to the Rasperry Pi. How does JMRI understand what `PT:XYZ1234:21:2:12` means?

Or do you mean that another developer later will add code for the device `XYZ1234`?

Reply:
JRMI doesn't understand :XYZ1234:21:2:12.  Or, for that matter, :MCP23017:1:38:6.  What JMRI does understand is that the string between the first two colons (XYZ1234 or MCP23017) should lead to a class "Provision<Type>" (ProvisionXYZ1234, ProvisionMCP23017).  That class is loaded dynamically (using class.forname).  (I haven't worked out the details, but it's my intent that the class could be added by the user, without any involvement from JMRI developers.)

In any case, everything after the second colon (21:2:12 or A,3,X) is interpreted by the appropriate Provision class and the rest of JMRI knows nothing specific about it.

Also, slightly different, but perhaps related...  The JMRI Raspberry Pi support depends on version 1.2 of Pi4j, but that's no longer being developed/supported.  Pi4j has transitioned to version 2...and version 2 doesn't include native support for GPIO extension devices.  They've also moved to a "plugin" model for extensions (though I haven't researched how their plugins work).  This may not affect JMRI as we have the full version 1.2 Pi4j but might become an issue down the road if we need to support new versions of the Raspberry Pi.

Reply:
> Also, slightly different, but perhaps related... The JMRI Raspberry Pi support depends on version 1.2 of Pi4j, but that's no longer being developed/supported. Pi4j has transitioned to version 2...and version 2 doesn't include native support for GPIO extension devices. They've also moved to a "plugin" model for extensions (though I haven't researched how their plugins work). This may not affect JMRI as we have the full version 1.2 Pi4j but might become an issue down the road if we need to support new versions of the Raspberry Pi.

It is probably advisable to update JMRI to use the newer version of the library now.

Relying on a library that is out of date for this work you are doing may result in more issues when the dependency is updated later.

Reply:
> It is probably advisable to update JMRI to use the newer version of the library now.

Well, yeah.  Though that's not what I had in mind when I began working on the GPIO extensions and I'm not sure I have the bandwidth to do that work in the next few months.  But I can at least take a look at the new version and maybe get some idea of the work involved in converting.

Reply:
On updating Pi4j...

First quick look.  Pi4j says they rely on JRE 11; we're running 8 (well, I'm running OpenJDK 1.8.0_212).  I added OpenJDK 11 to my RPi but ran into problems with it.  And I couldn't attach the debugger from my Windows system, so I'm not sure of the details but it appears that there are changes to reflection (which my new code is using.)

Reply:
A little more looking at Pi4j v2.  There are a few examples of plug-in code such as would be needed for MCP23017 or other GPIO extensions, but the examples are outside of the Pi4j project itself, appear to be user generated, and appear to be a long way from production quality.  If we move to Pi4j v2, and want to support extensions, it appears that we'd have to do our own development of the appropriate Pi4j plugins.

While looking in to this I also came across some of the reasoning behind their decision to remove extensions from the core project.  Their major concern had to do with long term support - folks would create an extension and add it to Pi4j but wouldn't necessarily be around to maintain it if issues arose down the road.

A problem perhaps not unfamiliar to the JMRI developer community.